installplan_operator/main.py

```diff
@@ -0,0 +1,229 @@
+
+from installplan_operator import config
+
+LOG = logging.getLogger(__name__)
+
+
+def create_openshift_client():
+    try:
+        kubernetes.config.load_incluster_config()
+    except kube_exc.ConfigException:
+        kubernetes.config.load_kube_config()
+
+    k8s = kubernetes.client.ApiClient()
+    return DynamicClient(k8s)
+
+
+class SubscriptionWatcher(threading.Thread):
+    '''Watch for events on Subscription resources'''
+
+    def __init__(self, q, api):
+        self.q = q
+        self.api = api
+        super().__init__(daemon=True)
+
+    def run(self):
+        for event in self.api.subscriptions.watch():
+            self.q.put(('subscription', event))
+
+
+class FileWatcher(threading.Thread):
+    '''Watch for filesystem events in config_dir'''
+
+    def __init__(self, q):
+        self.q = q
+        super().__init__(daemon=True)
+
+    def run(self):
+        for event in watch(config.config_dir):
+            self.q.put(('fs', event))
+
+
+class Approver:
+    def __init__(self):
+        self.oc = create_openshift_client()
+        self.q = queue.Queue()
+
+    def start_watchers(self):
+        self.tasks = [
+            FileWatcher(self.q),
+            SubscriptionWatcher(self.q, self),
+        ]
+
+        for task in self.tasks:
+            task.start()
+
+    @cached_property
+    def subscriptions(self):
+        '''This is an API endpoint for listing/getting/watching subscriptions'''
+
+        return self.oc.resources.get(
+            api_version='operators.coreos.com/v1alpha1',
+            kind='Subscription'
+        )
+
+    @cached_property
+    def installplans(self):
+        '''This is an API endpoint for listing/getting/watching installplans'''
+
+        return self.oc.resources.get(
+            api_version='operators.coreos.com/v1alpha1',
+            kind='InstallPlan'
+        )
+
+    def events(self):
+        '''An iterator that yields events from the queue.
+
+        This method will synthesize a timeout event if no events
+        are received from the queue after max_interval seconds.
+        '''
+
+        while True:
+            # Wait up to max_interval seconds for an event
+            try:
+                event = self.q.get(timeout=config.max_interval)
+            except queue.Empty:
+                event = ('timeout',)
+
+            yield event
+
+    def loop(self):
+        self.start_watchers()
+
+        t_last = 0
+        for event in self.events():
+            t_start = time.time()
+            t_delta = t_start - t_last
+
+            # Ignore triggers if we last checked less than min_interval
+            # seconds ago.
+            if t_delta < config.min_interval:
+                LOG.debug('ignoring %s event (too soon), t_last=%s, t_start=%s, t_delta=%s',
+                          event[0],
+                          t_last,
+                          t_start,
+                          t_delta)
+                continue
+
```

Comment:
Shouldn't you be also checking for the `('timeout',)` event from line 100 in here? Otherwise you will start processing phantom events, or have I missed something?
 
https://github.com/operate-first/installplan-operator/blob/95f760338fa843bca5b0232d160d109073b09899/installplan_operator/main.py#L100

Reply:
The reason the `timeout` event exists is so that we can run occasionally in the absence of events -- e.g., if we miss a legitimate event because it came in during the cooldown period from a previous trigger. This means that at least once every `config.max_interval` seconds we check to see if we need to do anything.

Reply:
I see, that make sense. So the timeout event is basically a pooling interval making the operator to force the parsing of the configs every 15. Ok, thanks.